src/jitlog/generator.lua

```diff
@@ -0,0 +1,358 @@
+      end
+
+      -- Was a length field specified
+      if not length then
+        if def.use_msgsize == t.name then
+          t.buflen = "msgsize"
+        elseif ftype == "string" then
+          length = t.name.. "_length"
+          assert(not fieldlookup[length], "cannot add automatic field length because the name is already taken")
+          add_field({name = length, noarg = true, type = "uint32_t"})
+          t.buflen = length
+          t.implicitlen = true
+        else
+          error("variable length field "..t.name.."did not specify a length field")
+        end
+      end
+    else
+      bitsize = self.types[ftype].size * 8
+    end
+
+    if bitsize < 32 then
+       --Check if we can pack into the spare bits of the id field
+      if (idsize + bitsize) <= 32 then
+        t.bitstorage = "msgid"
+        t.bitofs = idsize
+        t.bitsize = bitsize
+        idsize = idsize + bitsize
+      elseif t.type == "bitfield" then
+        offset = offset + bitsize
+      end
+    else
+      offset = 0
+    end
+  end
+
+  if vcount > 0 then
+    assert(not def.use_msgsize or vcount == 1, "Can't use msgsize for field size with more then one variable sized field")
+
+    for i, f in ipairs(vlen_fields) do
+      if f.buflen and not def.use_msgsize then
+        local buflen_field = fieldlookup[f.buflen]
+        buflen_field.lengthof = f.name
+        assert(buflen_field)
+      end
+    end
+    
+    sizefield = "msgsize"
+    --Add the implict message size field thats always after the msgid header
+    add_field({name = "msgsize", sizefield = true, noarg = true, type = "u32", writer = "vtotal"}, 2)
+  end
+  
+  local msgsize = 0
+  
+  for i, f in ipairs(fieldlist) do
+    local size = 0
+    local type = self.types[f.type]
+    if f.type == "bitfield" or f.bitstorage then
+      assert(f.bitstorage and fieldlookup[f.bitstorage])
+    elseif f.vlen then
+      assert(f.buflen)
+      assert(fieldlookup[f.buflen])
+    else
+      assert(type, "unexpected type")
+      size = type.size
+      if size < 4 then
+        
+      end
+      f.offset = msgsize
+    end
+    f.order = i
+    msgsize = msgsize + size
+  end
+
+  local result = {name = def.name, fields = fieldlist, vlen_fields = vlen_fields, fieldlookup = fieldlookup, size = msgsize, idsize = idsize, vsize = vcount > 0, 
+                  vcount = vcount, sizefield = "msgsize"}
+  return setmetatable(result, {__index = def})
+end
+
+function parser:parse_msglist(msgs)
+  for i, def in ipairs(msgs) do
+    local name = def.name
+    self:log("Parsing:", name)
+    local msg = self:parse_msg(def)
+    self.msglookup[name] = msg
+    table.insert(self.msglist, msg)
+  end
+end
+
+parser.builtin_msgorder = {
+  header = 0,
+}
+
+local function sortmsglist(msglist, msgorder)
+    local names = map(msglist, function(def) return def.name end)
+    msgorder = msgorder or {}
+    
+    -- Order the fixed built-in messages with a well know order
```

Comment:
I think we can just remove this comment (I guessed what `sort` does :) ).

Reply:
Maybe i should add that the reader expects some messages to have certain message id. Message type ids are assigned based on declaration order until recently the message list was passed in as hashtable which was terrible idea in retrospective

Reply:
What does sorting have to do with message IDs? That seems like a slightly orthogonal issue?

Reply:
The list of names is used to generate an enum for the message type

Reply:
I'm still baffled.